tlv/src/tlv_record.rs

```diff
@@ -0,0 +1,132 @@
+        let (length, consumed) = short_vec::decode_shortu16_len(buffer).ok()?;
+        if length > MAX_VALUE_LENGTH {
+            return None;
+        }
+        buffer.advance(consumed);
+        if length > buffer.len() {
+            return None;
+        }
+        Some(Self {
+            typ,
+            value: buffer.split_to(length),
+        })
+    }
+}
+
+/// Helper to allow serialization into BytesMut.
+struct BytesMutWriter<'a>(&'a mut BytesMut);
+
+impl Write for BytesMutWriter<'_> {
+    fn write(&mut self, data: &[u8]) -> io::Result<usize> {
+        self.0.extend_from_slice(data);
+        Ok(data.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl TlvSerialize for TlvRecord {
+    /// Write the TlvRecord instance into provided byte buffer.
+    fn serialize(&self, buffer: &mut BytesMut) -> Result<(), TlvEncodeError> {
+        // check for sanity
+        let opts = tlv_bincode_options();
+        if self.value.len() > MAX_VALUE_LENGTH {
+            return Err(TlvEncodeError::MalformedTlv);
+        }
+        let len = ShortU16(self.value.len() as u16);
+        #[allow(clippy::arithmetic_side_effects)] // this has no chance to overflow u64
+        let total_len = 1 + opts.serialized_size(&len).unwrap_or(3) + self.value.len() as u64;
+        if total_len > buffer.spare_capacity_mut().len() as u64 {
+            return Err(TlvEncodeError::NotEnoughSpace);
+        }
+        // write type
+        buffer.put_u8(self.typ);
+        // write length
+        let writer = BytesMutWriter(buffer);
+        let opts = tlv_bincode_options();
+        opts.serialize_into(writer, &len)?;
+        // write value
+        buffer.put_slice(&self.value);
+        Ok(())
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum TlvDecodeError {
+    #[error("Unknown type: {0}")]
+    UnknownType(u8),
+    #[error("Malformed payload: {0}")]
+    MalformedPayload(#[from] bincode::Error),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum TlvEncodeError {
+    #[error("Not enough space for payload")]
+    NotEnoughSpace,
+    #[error("Malformed TLV record")]
+    MalformedTlv,
+    #[error("Malformed payload: {0}")]
+    MalformedPayload(#[from] bincode::Error),
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        crate::{TlvEncodeError, TlvRecord, TlvSerialize},
+        bytes::{Bytes, BytesMut},
+    };
+
+    #[test]
+    fn test_serialize() {
+        let rec = TlvRecord {
+            typ: 1,
+            value: Bytes::from_static(b"blabla"),
+        };
+        let mut buffer = BytesMut::with_capacity(4);
+        assert!(matches!(
+            rec.serialize(&mut buffer),
+            Err(TlvEncodeError::NotEnoughSpace)
+        ));
+        let mut buffer = BytesMut::with_capacity(16);
+        rec.serialize(&mut buffer).unwrap();
+        assert_eq!(buffer[0], 1);
+        assert_eq!(buffer[1], 6);
+
+        assert_eq!(buffer[2..8], *b"blabla");
```

Comment:
nit: can we add a data structure serialize and deserialize and maybe cover more error types please?

Reply:
This is testing the TLV pet, TLV itself does not care about payload contents/type. We have tests for the application specific types in lib.rs